Mergeability rules self validation

lukasz-golebiewski opened the pull request:
## Overview
Follow-up for https://github.com/rchain/rchain/pull/2642


### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3638


### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

lukasz-golebiewski committed:
Add Polarity to Rho

lukasz-golebiewski committed:
Add Cardinality to Rho

lukasz-golebiewski committed:
Move Rho definition closer to it's instances

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    "!X !X"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S0, Nil),
+    "!X !4"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, F1),
+    "!X (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S0, F_)(Nil),
+    "!X !C"     -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, S1, C1),
```
We should avoid the $FOO.desc -> $FOO.mergable/conflicting duplication on each row by having the values in the map be instances of the follwing:

```scala
trait MergeabilityTestCase {
  val description: String
  def preconditions: F[Unit]
  def assertion: F[Unit]
}
```

then:
```scala
type Name = String // for lack of a better name, e.g. !4 !4

val mergeabilityCases = Map[Tag, MergeabilityTestCase](
  "!X !X"     -> SamePolarityMerge(S0, S0, Nil)
  ...
)
```
and the test would be sth. like:

```scala
  mergeabilityCases.toList.parTraverse { case (name, c) =>
    val result = (for {
      _ <- c.precondition
      _ <- c.assertion
    } yield ()).attempt
	result.map(name -> _)
  } >>= (results => /* summarize results into a nice assertion */)
```

To handle the MERGES/CONFLICTS subcases of COULD_MATCH_SAME_*, we can either:
 - have `if`-s in the implementation cases
 - have separate trait instances for the subcases (I prefer this)

ArturGajowy replied:
We can skip (get rid of) the `coveredBy` cases, or provide actual examples for them too.

lukasz-golebiewski replied:
What you propose isn't going to work because a `Map` cannot have duplicate keys and so the last entry for `!X !X` is going to win and all previous values are going to be ignored. Originally the description was used to disambiguate between the keys/tags/names. Another way to mitigate this would be to get rid of the `Map` and store tuples in a `List`. The second element of the tuple would be a sequence of cases (of size one for most). WDYT?

lukasz-golebiewski replied:
I am not in favor of introducing a single trait for all test cases. Implementing what I proposed in the comment above feels enough. I agree that it makes sense to split the `CouldMatch*` case into two.

ArturGajowy replied:
Good point! Using a map will cause 'duplicate key silent errors' without means of validating it. I'm against nested lists though. 

We should use a (flat) list, and enforce the names to be unique / make them unique by combining with the list index.

ArturGajowy replied:
> I am not in favor of introducing a single trait for all test cases.

It's more about the duplication of
`!X !4 -> $FOO.desc -> $FOO.{merges/fails}`
instead of having a common interface. The latter is just a means.

I think you'll find the trait needed to obtain the description for the assertion message. Notice also: once you change the `merges/conflicts` to `assertion` (which is needed, as the assertion should be same for all examples of a given case), the common interface is already there. Adding a trait atop is a matter of a couple `extends`.

lukasz-golebiewski replied:
This has a downside of unnecessarily complicating the test which checks whether all cases from the spreadsheet are covered. Hence I would rather go with a list. Unless there is a better argument for what you are proposing

ArturGajowy replied:
I'm fine if we fuse `preconditions` and `assertion` into just `assertion`

ArturGajowy replied:
> This has a downside of unnecessarily complicating the test which checks whether all cases from the spreadsheet are covered.

OTOH having the a list for each case instead of just a single value massively complicates the main test, and puts a lot of strain on the compiler - just to have 4-ish names have two tests instead of one.

The complication you mention is simply:
add the index before doing any set operations (e.g. all cases are one of those expected),
drop the index before doing any exhaustivity checks (e.g. each expected case has at least one representant)

ArturGajowy replied:
Frankly, you'd have coded it up faster than the time we've spent discussing this.

lukasz-golebiewski replied:
> massively complicates the main test, and puts a lot of strain on the compiler 

This is greately exaggerated

> Frankly, you'd have coded it up faster than the time we've spent discussing this.

The comment above could be spared, it doesn't bring anything. We're talking about sensibility of the approaches

Why are you suggesting combining with an index at all? Entries in a `List` do not have to be unique

ArturGajowy replied:
> This is greately exaggerated

Maybe w.r.t. to the compiler part. But usage-wise, reading and editing
`"4! 4!" -> List(CouldMatchSameMerges(X, Y, Z))`
on each line, is going to be a pain. It also makes the actual validations and the main test harder, b/c now you're handling a list of lists.

You're right that combining with an index might not even be needed. I suggested it as a means of regaining uniqueness, if it's needed in validations (e.g. we build a set somewhere). Most likely it's not needed though.

> The comment above could be spared, it doesn't bring anything. We're talking about sensibility of the approaches

I though really believe that the raised problem of `unnecessarily complicating the (spreadsheet coverage) test` would've been cheaper to handle in that test without this discussion. So while I'm all for considering sensibility of approaches, I hate to discuss such minutia.
I don't know, maybe I'm overlooking sth big in there?

lukasz-golebiewski replied:
```
Maybe w.r.t. to the compiler part. But usage-wise, reading and editing
"4! 4!" -> List(CouldMatchSameMerges(X, Y, Z))
on each line, is going to be a pain. It also makes the actual validations and the main test harder, b/c now you're handling a list of lists.
```

Maybe the reason I don't find this harder is that I'm used to handling nested lists and flattening doesn't give me headaches. I find grouping related cases under a single key more natural (and less error prone as it is easier to spot a missing case). Since this is largely opinion based I think we need another opinion to decide the way forward.

> I don't know, maybe I'm overlooking sth big in there?

Yes, the comment is inappropriate because perceived speed of developing a review suggestion doesn't say anything about it's sensibility

ArturGajowy replied:
I was asking about the complications in `unnecessarily complicating the (spreadsheet coverage) test`.

Re: the meta issue - sorry, I should've phrased that differently.

lukasz-golebiewski replied:
Looks like the `description` field is unnecessary. It's contents are going to the scaladoc

lukasz-golebiewski replied:
And since the precondition can also go away, so can the whole `MergeabilityTestCase` trait. `Effect[Unit]` being the second element of the tuple is enough

ArturGajowy replied:
I thought you'd say that :) Though to validate that the Name matches the MergeabilityCase we need a common interface to provide the name to - assuming we want to retain alignment and readability. To elaborate - we can either do:

`"!4 !4 -> HadItsMatch(X, Y, Z)` and in the test:
`(mergeabilityCase.checkMatchesName(name) >> mergeabilityCase.assertion).attempt`

or

`hadItsMatch("!4 !4", X, Y, Z)` and in the test:
`value.attempt`

I prefer the former, as the latter reads worse and aligns worse

lukasz-golebiewski replied:
Makes sense, although it's going to need to wait

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    "!X C!"     -> IncomingCouldMatch.desc      -> IncomingCouldMatch.conflicting(S0, C_, S0),
+    "!X (C!)"   -> PersistentCouldMatch.desc    -> PersistentCouldMatch.conflicting(S0)(C_, S0)(Nil),
+    "!4 !4"     -> CouldMatchSame.descConflicts -> CouldMatchSame.conflicting(S0, S1, F_),
+    "!4 !4"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, F1),
```
Notice that in:

`"!4 !4"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, F1),`

There's at least 5 kinds of errors possible:
1) mismatch of mergeability category between 1st and 2nd column
2) mismatch of mergeability category between 2nd and 3rd column (due to $FOO.{desc, merges} duplication)
3) mismatch of descMerges/descConflicts with 1st column
4) mismatch of mergable/conflicts with 2nd column
5) mismatch of actual test case with 1st column

The approach I suggest in previous comment elliminates 2) and 4) and will provide way of extension to elliminate the others too.

I'm pointing this out, because I find this crucial: whenever we pass two values separately (vs as a single value) we need to make sure the values are valid w.r.t each other, which requires separate logic outside the place we pass the values in.

ArturGajowy reviewed (changes requested):
Sharing what I have now, and proceeding to add more comments :)

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    "!4 !4"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, F1),
+    "!4 (!4)"   -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0)(S1, F_)(F0),
+    "(!4) (!4)" -> VolatileEvent.desc           -> VolatileEvent.mergeable(S0, F_)(S0, F_)(Nil),
+    "!4 !C"     -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, S1)(F0, C1),
```
The number of variants of arities is confusing and inconsistent across the cases.
E.g. here it should say `CouldMatchSame(S0)(S1)(F0, C1)` to be consistent
 
Let's have all MergeabilityCases take 3 curried parameters: left, right, base.
The last one can be vararg.

For 'volatile/local' cases we can have all of them be wildcards.

```scala
def samePolarityMerge(left: Rho)(right: Rho)(base: Rho*): MergrabilityTestCase = new ...

def volatileEvent(lefts: Rho*)(rights: Rho*)(base: Rho*): MergrabilityTestCase = new ...
```

This works with an `apply` method on a companion object as well.

Assertions underneath then become
e.g. 
```
//all are linear
assert((lefts ++ rights).forall(_.cardinality == Linear))
//same polarity
assert(left.polarity == right.polarity) 
//different polarities within one side 
assert(noDuplicatesBy(lefts)(_.polarity))
```

lukasz-golebiewski replied:
Good idea, I like it

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,31 +66,103 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      """
         |Two incoming sends/receives, at most one had a matching dual in TS.
         |The incoming events won't cause more COMMs together (same polarity).
         |They couldn't be competing for the same linear receive/send (at most one had a match).
```
consider this:

```
def matches(x: Rho, y: Rho) = {
  val (a, b) = if (x.polarity == Send) (x, y) else (y, x)
  val wildcards = List(F_, C_)
  val knownMatches = 
    List(S0, R0).flatMap(s => (F0 :: C0 :: wildcards).map(r => s -> r)) ++
    List(S1, R1).flatMap(s => (F1 :: C1 :: wildcards).map(r => s -> r))
  
 knownMatches.contains(a -> b)
}
```

Now we can incorporate the preconditions based on what matches what

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -99,149 +172,236 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    "!4 !!X"    -> SamePolarityMerge.desc       -> SamePolarityMerge.mergeable(S0, R1, F0),
+    "!4 !!4"    -> CouldMatchSame.descConflicts -> CouldMatchSame.conflicting(S0, R1, F_),
+    "!4 !!4"    -> CouldMatchSame.descMerges    -> CouldMatchSame.mergeable(S0, R1)(F0, F1),
+    "!4 CX"     -> HadItsMatch.desc             -> HadItsMatch.mergeable(S0, C_, F_),
```
I know you consider this overkill, but validating the name (`!4 CX`) matches the mergeability case (HadItsMatch) is what was missing for spotting quite a few wrong tests during the review automatically, and who knows how many currently and in the future.

At the same time, parsing the:
```
case class Event(p: Polarity, c: Cardinality) 
case class Example(leftIn: Event, rightIn: Event, leftBase: Option[Event], rightBase: Option[Event])
```
from the Name-s should really be straightforward.

Just as the assertions to be done based on them:
```scala
//same/different polarity:
ex.leftIn.polarity ==/!= ex.rightIn.polarity

//number of matches
(leftBase, rightBase).exists(_.isEmpty) //at most one match
```

I implore you to reconsider this :)

lukasz-golebiewski replied:
Ok, I'm going to reconsider

lukasz-golebiewski replied:
This will have to wait until after rules for peek are implemented

ArturGajowy replied:
I can live with that. Can you though? :) I mean, that's additional safety net that would help both of us when implementing peek. I'm fine either way though :)

lukasz-golebiewski replied:
I treat this self validation as a means to spare the future maintainers some stress. I've been looking at those examples for so long that I know them by heart already, so yes, I can live with that

lukasz-golebiewski committed:
Remove comments

Improve formatting

lukasz-golebiewski committed:
Do not store mergeability cases in a Map

This is to remove the probability of cases not being tested because
they are shadowed by an entry with the same key

lukasz-golebiewski committed:
Extract SamePolarityMerge object

lukasz-golebiewski committed:
Extract CouldMatchSameConflicts

lukasz-golebiewski committed:
Extract CouldMatchSameMerges

lukasz-golebiewski committed:
Introduce base mergeable and conflicting case

lukasz-golebiewski committed:
Extract IncomingCouldMatch

lukasz-golebiewski committed:
Extract VolatileEvent

lukasz-golebiewski committed:
Extract PersistentCouldMatch

lukasz-golebiewski committed:
Extract PersistentNoMatch

lukasz-golebiewski committed:
Fix erroneously labeled cases

lukasz-golebiewski committed:
Remove description from the mergeability cases list key

lukasz-golebiewski committed:
Move cases' descriptions to scaladoc

lukasz-golebiewski committed:
Do not use Rho.`|` in tests' definitions

lukasz-golebiewski committed:
Add polarity/cardinality checks

lukasz-golebiewski committed:
Add match checks

dzajkowski commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    val (a, b)    = if (x.maybePolarity == Some(Send)) (x, y) else (y, x)
+    val wildcards = List(F_, C_)
+    val knownMatches =
+      List(S0, R0).flatMap(s => (F0 :: C0 :: wildcards).map(r => s   -> r)) ++
```
`.map(r => s   -> r)) ++` formatting?

lukasz-golebiewski replied:
This was auto-formatted, sorry

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -204,23 +204,19 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
     "!C (C!)"   -> PERSISTENT_COULD_NOT_MATCH -> coveredBy("!C (!C)"),
     "4X 4X"     -> SAME_POLARITY_MERGE        -> merges(F_, F_, Nil),
     "4X 4!"     -> SAME_POLARITY_MERGE        -> merges(F0, F_, S1),
-    // Skipping 4X 4!! merges, 4X !!X may merge or not, 4X !!4 may merge or not
```
We should add those missing cases at some point though. Maybe in this PR?

lukasz-golebiewski replied:
No, this one is big enough already

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -131,6 +131,13 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+  object SamePolarityMerge {
+
+    def apply(left: Rho*)(right: Rho*)(base: Rho*): Effect[Unit] =
+      merges(left.reduce(_ | _), right.reduce(_ | _), base.reduce(_ | _))
```
Love the ASCII art :heart:

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -184,103 +184,102 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-          val key = s"$name $desc"
-          y.attempt.map(_.fold(_ => s"$key FAILED", _ => key))
+        case (key, test) =>
+          test.attempt.map(_.fold(_ => s"$key FAILED", _ => key))
```
How do we discern same-named cases when a test fails?
One way out would be to force distinct names :)

lukasz-golebiewski replied:
I would rather add (very) short descriptions to all test cases

ArturGajowy replied:
How about threading the source position?

ArturGajowy replied:
How about using the source position and displaying it in the assertion?

ArturGajowy replied:
(are we doing that already? If so - case closed for me :))

lukasz-golebiewski replied:
It's enough to revert back to:
```
it should "respect mergeability rules when merging blocks" in effectTest {
    mergeabilityCases.map(_._2).parSequence
  }
```

but then it will only display the first failure. I think this is good enough

lukasz-golebiewski replied:
This is already the case with precondition failures

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+  def differentPolarities(rho: Rho*)(
+      implicit
+      pos: source.Position
+  ): Any =
```
please add a better type :)

lukasz-golebiewski replied:
Like what?

ArturGajowy replied:
Like Seq[Assertion] f.e. Change the body if needed. See my latter comment on how you can change this.

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+  def samePolarities(left: Seq[Rho])(right: Seq[Rho])(
+      implicit
+      pos: source.Position
+  ) =
```
please add a return type :)

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      implicit
+      pos: source.Position
+  ): Any =
+    if (rho.size == 2) {
```
While technically it works, because there are only two polarities (or 3, if we count None, which doesn't affect the assertion) - it's quite confusing and raises doubt immediately.

I'd much prefer a solution that does `assert(noDuplicates(rho.flatMap(_.maybePolarity))`

lukasz-golebiewski replied:
Ok, I'll write something similar

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -110,35 +227,86 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      differentPolarities(left)(right)
+      crossedPolarity(left)(base)
+      crossedPolarity(right)(base)
```
I'm not sure about the relevance of the `crossedPolarity` check. Its occurrence isn't used in the reasoning in the scaladoc; it is only checked for cases where there is two terms in base deploy. I also think whatever it checks is going to be coverd by `someMatch` case in the latter commits (a match implies existence of different polarities). I'd remove it, but maybe there's a benefit I'm missing?

ArturGajowy replied:
Also, its definition is very special-cased

lukasz-golebiewski replied:
Yes, this was added before the "match" assertions and is now superfluous. It can go away

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      pos: source.Position
+  ) = {
+    assert(rho.forall(_.maybeCardinality.isDefined))
+    assert(rho.forall(_.maybeCardinality.get == Linear))
```
people and tools justly frown at the usage of `Option.get`. Why not just `assert(rho.forall(_.maybeCardinality == Some(Linear))`?

lukasz-golebiewski replied:
`.get` is acceptable in tests, but you're right, the error message is going to be friendlier when written like you proposed

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -110,35 +227,86 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      if (left.size == 1 && right.size == 1) fail("No volatile COMM event")
+      differentPolarities(left: _*)
+      allLinearWhenTwo(left)
```
I think this too is going to be covered by the `matches` cases added later. Can we remove it?

lukasz-golebiewski replied:
I'd leave this as it's an additional safety measure for not letting persistent operations into COMM events for this case

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      implicit
+      pos: source.Position
+  ) = {
+    assert(rho.forall(_.maybeCardinality.isDefined))
```
this is more than the function name says. It's also not relevant to the preconditions this method's used in. I'd remove this line.

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,14 +66,100 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      pos: source.Position
+  ) = {
+    assert(rho.forall(_.maybeCardinality.isDefined))
+    assert(rho.exists(_.maybeCardinality.get == NonLinear))
```
We can avoid the `.get` by `rho.flatMap(_.maybeCardinality).exists(_ == NonLinear)`

ArturGajowy replied:
Also, this (and what's currently in code) would be `atLeastOnePersistent`. We should probably just make it `rho.filter(_.maybeCardinality == Some(NonLinear)).size == 1`

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -234,14 +402,14 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
     "(!C) CX"   -> PersistentNoMatch(S0, C_)(C_)(Nil),
     "(4!) (4!)" -> VolatileEvent(F_, S0)(F_, S0)(Nil),
-    "(4!) (C!)" -> VolatileEvent(F_, S0)(C_, S0)(Nil),
+    "(4!) (C!)" -> PersistentNoMatch(F_, S0)(C_, S0)(Nil),
```
yay! :joy:

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -154,14 +174,38 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    (for {
+      l <- left
+      r <- right
+    } yield matches(l, r)).fold(true)(_ && _)
```
We could make that:
```
    Applicative[List]
      .product(left.toList, right.toList)
      .forall(Function.tupled(matches))
```

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -154,14 +174,38 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    (for {
+      l <- left
+      r <- right
+    } yield matches(l, r)).fold(false)(_ || _)
```
Similarly, this could be:
```
    Applicative[List]
      .product(left.toList, right.toList)
      .exists(Function.tupled(matches))
```

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -154,14 +174,38 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    } yield matches(l, r)).fold(false)(_ || _)
+
+  def ifCommEvent(rhos: Seq[Rho])(f: (Rho, Rho) => Assertion) =
+    if (rhos.size == 2) {
```
this doesn't check that the args form a comm event. Maybe this should be `ifTwo`?

lukasz-golebiewski replied:
Yes :-)

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -176,6 +220,7 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
       samePolarities(left)(right)
       differentPolarities(left)(base)
       differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base))
```
This would also pass if there were 2 matches, as long as there is one non-matching pair.
At this point, operating on lists for non-volatile cases really gets in the way.
We could say `assert(someMatch(left, base) != someMatch(right, base))`.
What we would really like to say, is:
`assert(hasMatch(left, base) != hasMatch(right, base))`, where left and right are singular Rho (not lists).

lukasz-golebiewski replied:
What you're suggesting isn't going to work with `"!X !X"     -> SamePolarityMerge(S0)(S0)(Nil),`

ArturGajowy replied:
The problem I raised still holds though. You get the idea. We should be asserting that at most one incoming event had a match.

lukasz-golebiewski replied:
Yeah, I'll fix this

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] =
+      super.precondition(left: _*)(right: _*)(base: _*) >> {
```
Can we replace the inheritance with calling a common method with a telling name?

lukasz-golebiewski replied:
If you insist

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base) || onePersistent(base))
```
This should be `allMatch(left, base) && allMatch(right, base`), + an assertion that they've either matched different events, or the matched-by-both event is non-linear.

I can't make a connection between the assertion in code ane the mergability rules/cases we're testing.

if we had signular `left` and `right`, this could be:

```
val m1 = findMatch(left, base) //throws asseriton error if not found
val m2 = findMatch(right, base) // we could also return Option-s and do the needed assertions
assert(m1 != m2 || m1.cardinality == NonLinear || m2.cardinality == NonLinear)
```

Notice the `Conflicts` case is exactly the same, just with the last assertion negated.
(`m1 == m2 && m1.cardinality == Linear && m2.cardinality == Linear`)

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+      differentPolarities(left)(right)
+      crossedPolarity(left)(base)
+      crossedPolarity(right)(base)
+      assert(someMatch(left, base) || someMatch(right, base))
```
it'd be great to encode the
```
 either:
   - both were linear
   - one was non-linear, the other had a match
```
criteria for the incoming events.

With singular left/right:
```
val (a, b) = if (left.cardinality == Linear) (left, right) else (right, left)
assert(a.cardinbality == Lineear && (b.cardinality == Linear || hasMatch(a, base)))
```

ArturGajowy commented on casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala:
```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
-        |The other deploy had an event without a match in TS, of same polarity to the non-linear.
-        |These could not spawn more work.
-        |""".stripMargin
+  trait TestCase {
```
Would be great to move this ant its descendants ot a separate file/package, probably with Rho and the util assertions

lukasz-golebiewski replied:
Ok, done

ArturGajowy reviewed (changes requested):
Great work! Had some questions/remarks

lukasz-golebiewski committed:
Address review comments

lukasz-golebiewski committed:
Extract mergeability rules to a separate trait

lukasz-golebiewski commented:
bors r+
